Software Engineer's guide to A/B testing

Lior539 opened the pull request:
Software Engineer's guide to A/B testing.

This blog was harder than expected to write, as I was trying to avoid making it a generic SEO blog. I was also trying to avoid too much repetition between this blog and the [A/B testing mistakes blog
](https://posthog.com/blog/ab-testing-mistakes), since there is a lot of overlap on the topics they discuss.

Looking for feedback on:
- content of the blog. Is it interesting and actionable?
- Is the blog too "heavy". It's quite long and a lot of text. Perhaps I need to add more images to break it up? (open to suggestions here).

[Issue](https://github.com/PostHog/posthog.com/issues/6219)

Lior539 commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,216 @@
+
+- **Accept the results:** Sometimes, an insignificant result is the correct result. It tells you that the change you made didn't have the impact you thought it would. This is still a valuable insight, and helps you shipping unnecessary changes.
+
+
+Calculate Effect Size
+Review Secondary Metrics
+
+how to interpret the results - a few examples 
+ - what happens if not statistical signfiance?
+ - what happens if statsical significant 
+determining your next steps
+
```
```suggestion

```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+author: ['lior-neu-ner']
+featuredImage: ../images/blog/experiment-hog.jpg
+featuredImageType: full
+featuredVideo: https://www.youtube-nocookie.com/embed/C2ICJbICfsg
```
```suggestion
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+A/B testing enables you to be confident that your product changes are having their intended effects. Rather than making decisions based on gut feeling or intuition, A/B testing provides empirical data to help you remove bias from decision making.
+
+Companies like [Netflix](https://uxdesign.cc/how-netflix-does-a-b-testing-87df9f9bf57c), [Uber](https://www.uber.com/en-GB/blog/supercharging-a-b-testing-at-uber/) and [Booking.com](https://booking.ai/tagged/ab-testing) use A/B testing to ship hundreds of changes a day. In this blog, we'll teach you basic concepts to get you started in A/B testing and supercharge your product growth!
```
```suggestion
Companies like [Netflix](https://uxdesign.cc/how-netflix-does-a-b-testing-87df9f9bf57c), [Uber](https://www.uber.com/en-GB/blog/supercharging-a-b-testing-at-uber/) and [Booking.com](https://booking.ai/tagged/ab-testing) use A/B testing to improve their apps in hundreds of ways each day. In this post, we'll teach you basic concepts to get you started in A/B testing and supercharge your product growth!
```
A/B tests are a bit downstream of just "shipping"

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+Companies like [Netflix](https://uxdesign.cc/how-netflix-does-a-b-testing-87df9f9bf57c), [Uber](https://www.uber.com/en-GB/blog/supercharging-a-b-testing-at-uber/) and [Booking.com](https://booking.ai/tagged/ab-testing) use A/B testing to ship hundreds of changes a day. In this blog, we'll teach you basic concepts to get you started in A/B testing and supercharge your product growth!
+
+> **Note:** A/B tests are commonly referred to as "experiments".
```
```suggestion
```
Lets just mention this when it comes up

andyvan-ph replied:
Agreed

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+## How A/B testing works
+
+In A/B testing, you first decide on a goal metric and what change you'd like to see. For example, you may to increase conversion rate, or you may want to decrease customer churn rate. Then, once you've made your code changes and are ready to test, you **randomly** split your users into two (or more) groups:
```
Have an idea of something you'd like to change or test might come first?

ivanagas replied:
```suggestion
In A/B testing, you first decide on a goal metric and what change you'd like to see. For example, you may to increase conversion rate or decrease customer churn rate. Once you've made your code changes and are ready to test, you **randomly** split your users into two (or more) groups:
```
- Where changes take place is important to decide on too.
- How do you split your users up? Using a service? config?

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+There are five parts to a good A/B test:
+
+### 1. Has a clear, measurable goal
```
```suggestion
### 1. A clear, measurable goal
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+Like "increase conversion rate by 10%", or "decrease churn rate by 5%". Without a clear goal, it can be unclear what is a successful outcome of test.
+
+### 2. Has a clear hypothesis why your your changes will achieve your goal
```
```suggestion
### 2. A clear hypothesis why your your changes will achieve your goal
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+- The control group: these users will experience your product as is – without any of your new changes.
+- The test group: these users will experience your product with your new changes.
+
+You then run your A/B test for some time, usually a few days or weeks, to gather data and compare the differences in the goal metric between the different groups.
```
A little bit more on calculating how long could be valuable here.

Lior539 replied:
This is covered more below. But based on this (and some of your other feedback), I think it would be useful to outline what topics will be covered in the intro, so to give an idea of when to expect each topic

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+![How A/B testing works](../images/blog/ab-testing-guide-for-engineers/ab-test-illustration.png)
+
+> **Note:** These groups are also referred to as "variants".
```
Groups refers to users, variants usually refers to the code/config so I wouldn't say so. Maybe needs rewording.

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+This focuses your testing on what changes you're making and why.
+
+Here's an example of a good hypothesis for a SaaS Product: "Adding a short tutorial video on the login page will increase user engagement by 30% because it will help new users understand how to use our product more effectively."
```
```suggestion
Here's an example of a good hypothesis for a SaaS product: "Adding a short tutorial video on the login page will increase user engagement by 30% because it will help new users understand how to use our product more effectively."
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+### 3. Tests as small of a change as *reasonably* possible 
+
+Ideally, an A/B test should change one thing at a time. This ensures that any change in user behavior can be attributed to the variable that was changed. If multiple variables are changed, it's difficult to determine which one caused the change in behavior.
```
```suggestion
Ideally, an A/B test should change one thing at a time. This ensures that any change in user behavior can be attributed to what was changed. If there are multiple changes, it's difficult to determine which one caused the change in behavior.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+Ideally, an A/B test should change one thing at a time. This ensures that any change in user behavior can be attributed to the variable that was changed. If multiple variables are changed, it's difficult to determine which one caused the change in behavior.
+
+However, the caveat here is that too small of change can slow your team down. Since running A/B tests take time, if you're constantly testing small changes it may take long to ship large, meaningful changes.
```
```suggestion
The caveat here is that too small of change can slow your team down. Since running A/B tests take time, if you're constantly testing small changes, it may take long to ship large, meaningful changes.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+However, the caveat here is that too small of change can slow your team down. Since running A/B tests take time, if you're constantly testing small changes it may take long to ship large, meaningful changes.
+
+A good rule of thumb to know if your change is too small is if it's unlikely to impact user behavior significantly. To help inform this decision, you can use qualitative data from user research or quantitative data from previous A/B tests.
```
The right change size depends on:
- Scale of company (number of users involved in A/B tests). Big companies make small changes, small companies make big changes.
- Risk of change (potential impact). Changing a finance algorithm that can lose $ms shouldn't be A/B tests, but a homepage piece of copy can.

Impact user behavior significantly is too vague IMO.

ivanagas replied:
Also quantitative data doesn't have to be only from previous A/B tests, normal analytics can be used too.

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+A good rule of thumb to know if your change is too small is if it's unlikely to impact user behavior significantly. To help inform this decision, you can use qualitative data from user research or quantitative data from previous A/B tests.
+
+### 4. Has a sufficiently large sample size of users
```
```suggestion
### 4. A sufficiently large sample size of users
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+To know whether your change is large enough, you'll need to know:
+
+1. Your current conversion rate.
```
It's not necessarily just conversion, can be growth in other metrics as well.

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+2. Your "minimum detectable effect" i.e., the smallest change in the conversion rate that you want to detect. The smaller the change, the larger the sample size you'll need.
+3. Your desired level of confidence (the industry standard is 95%).
+
+You then use a [formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself (We also include this calculator when creating a new [A/B test in PostHog](/ab-testing/features)).
```
Can you show what the formula is here?

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+You then use a [formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself (We also include this calculator when creating a new [A/B test in PostHog](/ab-testing/features)).
+
+### 5. Has a long enough test duration
```
```suggestion
### 5. A long enough test duration
```

ivanagas replied:
Aren't sample size and duration linked? Could this be one point?

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+Once again, there are many calculators online to help you determine your duration, and PostHog also includes this calculator when creating a new experiment.
+
+**A good rule of thumb is that your experiment duration should be between one week and one month:
+**
```
```suggestion
**A good rule of thumb is that your experiment duration should be between one week and one month:**
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+**A good rule of thumb is that your experiment duration should be between one week and one month:
+**
+- One week is a good minimum since users may behave differently on weekends or weekdays.
+- One month is a good maximum since otherwise you may delay shipping important product changes.
```
How long do we normally run A/B tests for? What about top companies?

This could be a graphic.

Lior539 replied:
It always depends on what you're testing, your sample size, and the expected change in metric. Its differs wildy across features (not companies). I dont think its worth including

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+## When not to A/B test
+
+There's are costs and trade-offs to running A/B tests, and so it may not be a good idea to always run them. Here a few scenarios in which you may not want to run an A/B test:
```
```suggestion
There's are costs and trade-offs to running A/B tests so it may not always be a good idea to run them. Here a few scenarios in which you may not want to run an A/B test:
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+
+- **When you lack of sufficient traffic or users, or have time constraints:** It won't be possible to gather enough data to obtain the statistically significant results required to make good decisions.
+
+- **When you have high implementation costs:** Sometimes it may create a lot of technical debt to support and maintain both control and test variants of your experiment.
+
+- **When you have opportunity costs:** The time and resources it takes to implement and analyze A/B tests may be better spent working on other projects.
```
These could be combined?

Lior539 replied:
Hmm perhaps. I think better to have them separate. Opprotunity cost is implied in "high implementation costs". but I think it's better to call it out as it's own point (it's not always necessarily related to implementation)

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,207 @@
+- One week is a good minimum since users may behave differently on weekends or weekdays.
+- One month is a good maximum since otherwise you may delay shipping important product changes.
+
+## When not to A/B test
```
This doesn't feel like a super compelling section right now. Feels a bit like "when an A/B test is the opposite of a good A/B test," so not super necessary.

andyvan-ph replied:
I kind of agree, mainly because this breaks the flow somewhat to the next section, but I think it's still useful info. Maybe we move this into the FAQ section?

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
````diff
@@ -0,0 +1,207 @@
+
+To do so, ensure that checking your feature flag and logging their exposure is the *absolute* last condition you check in your code: 
+
+```
````
Could we do a multi language block here? or would that look dumb?

Lior539 replied:
This is more pseudo-code than actual code, so I dont think we need to add a multilanguage block

ivanagas replied:
I'm thinking like one tab of multilanguage block says "Incorrect" and the other says "Correct" instead of having one long block.

Lior539 replied:
ah, nice idea. I'll try it out

Lior539 replied:
Tried it out but didnt render correctly. It rendered as 2 separate code blocks. I dont think the multilangauge compeonet is enable in the blog?

ivanagas reviewed (commented):
Some edits and changes needed.

Important to think about: is this a newsletter issue? or just a blog?

If its a newsletter issue, I think it needs to better match the tone/style of a newsletter: https://newsletter.posthog.com/p/6-mistakes-weve-made-at-posthog-and. Also needs to be cut down significantly.

If it's a blog, what will get people to read this over what we already have and what others have written? Is this HackerNews worthy? SEO?

Lior539 commented:
> Important to think about: is this a newsletter issue? or just a blog?

I had a blog in mind for this

> Is this HackerNews worthy? SEO?

I dont think its HN worthy (yet), and I wouldnt say the main goal is SEO. It's more for having completeness of educational resources around this topic. If our goal to be the hub for product eng to find resources, part of is making sure we have all the resources they need, so that when they decide they need to learn about a topic, they know they'll be able to find it on PostHog

Rereading @andyvan-ph's [original issue](https://github.com/PostHog/posthog.com/issues/6212), I realize he did intend for it to be a newsletter issue. Im happy to convert it if we feel its a good idea

> If it's a blog, what will get people to read this over what we already have and what others have written? 

The idea is that this is more software engineer focused (or at least thats what Im trying to achieve). Most the other resources on A/B testing out there target marketers or other personas. There's nothing thats quite SWE focused

ivanagas commented:
Stepping on toes a bit here, but I'm afraid that if we just do things for "completeness," people won't read them. We should try to write things that fit into one of the buckets [here](https://posthog.com/handbook/growth/marketing#content)

<img width="774" alt="Screen Shot 2023-07-06 at 10 32 13 AM" src="https://github.com/PostHog/posthog.com/assets/34755028/3060e82a-8589-4d58-aab1-746bbd104682">

If you can, I think you should try to convert this into a newsletter, but it might require a big rewrite.

andyvan-ph commented:
I think this is a good first draft, it's 80% there. Some feedback:

- The basic format is fine. I don't expect this to be big SEO piece (it's very competitive), but it should be a potentially shareable resource – something people will want to bookmark. This post will form part of the newsletter, but it shouldn't BE the newsletter so we're good there.

- I feel like the intro + "How A/B testing works" aren't right tonally. They sound like we're ELI5-ing A/B testing, but I'd expect most software engineers to at least be familiar with the core concepts. That's not to say we shouldn't explain the basics, but I think we go a teeny bit deeper – I appreciate this is a fine line to tread!

- The 'How A/B testing works' ought to be a little more technical – can we introduce different types of testing, such as multi-armed bandit? I think we can afford to make this section a bit longer. I might be wrong here, though.

- We use examples like "increase conversion rate" or "increase signups" throughout. I'd prefer we use examples that feel more product orientated, rather than growth orientated, if that makes sense?

- I think this structure is sort of there, but maybe we should lean into this as guiding someone through running an A/B test from beginning to end. So it could be something like this:
  1. Deciding whether to run an A/B test – i.e. can you get a large enough sample size, do you have time, when not to run one etc.
  2. Creating your goal and / or hypothesis
  3. Implementing your A/B test
  4. Monitoring your A/B test
  5. Analyzing your results

- If we use the above structure, I think it would be nice to finish with a summary and (maybe) provide an A/B testing checklist people can use. This way, the reader goes away with something they can use / share, and they might come back to the guide as a refresher

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,205 @@
+---
+date: 2023-07-04
+title: "A software engineer's guide to A/B testing"
```
Suggestion:

- An engineer's guide to running successful A/B tests

andyvan-ph replied:
Some variations...

- A/B tests that work: A software engineer's guide
- A software engineer's guide to A/B tests that work

You might be right about your original.

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,205 @@
+  - Product metrics
+---
+
+A/B testing enables you to be confident that your product changes are having their intended effects. Rather than making decisions based on gut feeling or intuition, A/B testing provides empirical data to help you remove bias from decision making.
+
+Companies like [Netflix](https://uxdesign.cc/how-netflix-does-a-b-testing-87df9f9bf57c), [Uber](https://www.uber.com/en-GB/blog/supercharging-a-b-testing-at-uber/) and [Booking.com](https://booking.ai/tagged/ab-testing) use A/B testing to ship hundreds of changes a day. In this post, we'll teach you basic concepts to get you started in A/B testing and supercharge your product growth!
```
I think we can come back to this last, but this intro doesn't work for me. Can we lean into your credentials here? Afaik, you have a lot of experience doing A/B tests at Meta etc. (and on your own project), let's talk about that. Write it in the first person ("I will guide you through") rather than third person ("we'll teach you the basics").

andyvan-ph replied:
Maybe have a play around with this in an intro and throughout. Where possible, it's better to be opinionated. Drawing from your experience, and writing in first person, might lift the whole piece up a level.

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,205 @@
+
+### 1. A clear, measurable goal
+
+Like "increase conversion rate by 10%", or "decrease churn rate by 5%". Without a clear goal, it can be unclear what is a successful outcome of test.
```
I'm not the expect here, but surely it's acceptable to have a metric you want to measure without a pre-defined "I want to increase it by this much" success goal?

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,205 @@
+
+### 2. A clear hypothesis why your your changes will achieve your goal
+
+This focuses your testing on what changes you're making and why.
+
+Here's an example of a good hypothesis for a SaaS product: "Adding a short tutorial video on the login page will increase user engagement by 30% because it will help new users understand how to use our product more effectively."
```
I think we should add a bit more here. Feels too light.

Lior539 replied:
Yeah fair enough. I was trying to avoid repeating what I said in the other blog, but I do need to expand more here

Lior539 commented:
> Stepping on toes a bit here, but I'm afraid that if we just do things for "completeness," people won't read them. We should try to write things that fit into one of the buckets [here](https://posthog.com/handbook/growth/marketing#content)

@ivanagas  Thanks for sharing this. As @andyvan-ph mentioned, I think it falls in the "shareable content" bucket. My goal with this is to create something that engineers will find genuinely helpful and informative, especially if they're new to A/B testing

> This post will form part of the newsletter, but it shouldn't BE the newsletter so we're good there.

@andyvan-ph what do you mean by "form part of the newsletter" ? How will it feature?.

Thanks for the rest of the feedback too. Will rework it!

Lior539 commented:
I've just added an update. I think it's in a better place now and feel the content is meaningful and interesting without being too heavy or generic

Please have a read through and let me know what you think. You dont need to give grammar or phrasing suggestions just yet. I'll clean it up if we think the blog is mostly there in terms of content.

> I feel like the intro + "How A/B testing works" aren't right tonally. They sound like we're ELI5-ing A/B testing, but I'd expect most software engineers to at least be familiar with the core concepts. That's not to say we shouldn't explain the basics, but I think we go a teeny bit deeper – I appreciate this is a fine line to tread!

Yeah this is super tricky. In my experience, most engineers have never run A/B tests. Their understanding is usually "splitting users into 2 groups and comparing results", but nothing really beyond that. So its tricky to avoid ELI5-ing. 

I went back and forth on this a ton this morning, and Im not sure it's worth updating. I like how its succint and to the point, so that we can move on and talk about the other topics.

Now that I've updated it again, mind reading through it again and letting me know if we should still update?

> We use examples like "increase conversion rate" or "increase signups" throughout. I'd prefer we use examples that feel more product orientated, rather than growth orientated, if that makes sense?

I prefer to use these, since they're easy to understand (everyone's familiar with them) and you dont need to do any additional explaining. More "product" ones tend to be increasing feature usage or similar, which I dont think sounds as nice. I'm open to suggestions, though

andyvan-ph commented:
> @andyvan-ph what do you mean by "form part of the newsletter" ? How will it feature?.

Roughly speaking, we'll pull key bits from this guide and then direct newsletter readers to your full guide. A bit like how [this one](https://newsletter.posthog.com/p/what-weve-learned-about-product-market) is mostly a summation of James' larger guide.

andyvan-ph commented:
> Yeah this is super tricky. In my experience, most engineers have never run A/B tests. Their understanding is usually "splitting users into 2 groups and comparing results", but nothing really beyond that. So its tricky to avoid ELI5-ing.

All good. Might just be easier to say something along the lines of... "If you're already familiar, skip to xyz. If not, here's a quite ELI of A/B testing."

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,250 @@
+  - Product metrics
+---
+
+A/B testing enables you to be confident that your product changes are having their intended effects. Rather than making decisions based on gut feeling or intuition, A/B testing provides empirical data to help you remove bias from decision making.
+
+This post is for software engineers who are just getting started with A/B testing. We'll teach you the basic concepts on how to run a successful test and supercharge your product growth!
+
+We'll cover:
+- [How A/B testing works](#how-ab-testing-works)
+- [What makes a good A/B test?](#what-makes-a-good-ab-test)
+- [How to implement a good A/B test](#how-to-implement-a-good-ab-test)
+- [Monitoring your A/B test](#monitoring-your-ab-test)
+- [Analyzing your results](#analyzing-your-results)
+- [Frequently asked questions](#frequently-asked-questions)
+- [Launch checklist](#launch-checklist)
```
```suggestion
As a software engineer, you have two options:

1. YOLO every change and hope they have the desired impact.
2. Track user metrics and run A/B tests to verify your changes have the desired effect.

One is easier than other, but the easy way rarely works in the long term. This guide is for software engineers who are just getting started with A/B testing. In it, you'll learn

- How to devise good A/B tests
- How to implement and monitor your tests
- How to analyze you results (statistically significant, or not)

I'll also cover some frequently answered questions, such as when you _shouldn't_ run A/B tests, and there's a simple [A/B test launch checklist]((#launch-checklist)) you can use when you run your own tests.
```

Lior539 replied:
Love the new intro!

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,250 @@
+---
+date: 2023-07-04
```
```suggestion
date: 2023-10-04
```

andyvan-ph reviewed (commented):
Left a bunch of suggestions, particularly a revamp to the intro. Lottie's made a nice little graphic to with hedgehogs I need to add, so this is pretty close now.

Lior539 commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,252 @@
+
+Like "increase paid conversion rate", or "decrease churn rate". This should be a *single* metric that you're aiming to improve.
+
+Without a clear goal, it's unclear what is a successful outcome of test. For example, a bad goal is "increase user engagement", since it's unclear which metric will define success – is it daily active visitors, total number of page views, or time spent in the app?
```
```suggestion
Without a clear goal, it's unclear what a successful outcome of the test is. For example, a bad goal is "increase user engagement", since it's unclear which metric will define success – is it daily active visitors, total number of page views, or time spent in the app?
```

Lior539 commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,252 @@
+
+2. **Any secondary metrics.** These are usually metrics that are closely related to your goal metric. For example, if your goal metric is to increase paid conversions, a secondary metric may be to increase sign-ups.
+
+3. **Any counter metrics.** These are metrics that ensure your user experience isn't degrading. For example, if your goal is to increase sign-ups, you should also verify your churn rate is not increasing.
```
```suggestion
3. **Any counter metrics.** These are metrics that ensure your user experience isn't deteriorating. For example, if your goal is to increase sign-ups, you should also verify your churn rate is not increasing.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,250 @@
+
+The caveat here is that too small of change can slow your team down. Since running A/B tests take time, if you're constantly testing small changes, it may take long to ship large, meaningful changes.
+
+A good rule of thumb to know if your change is too small is if it's unlikely to impact user behavior significantly. To help inform this decision, you can use qualitative data from user research or quantitative data from existing logging or previous A/B tests.
```
I was referring to using qual / quant data to decide data to determine if a change is likely to impact user behavior. On reflection, it's not really a big deal. We shouldn't teach people to suck eggs!

Lior539 commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,250 @@
+
+**Before launch**
+
+- [ ] A clear goal metric.
+- [ ] A clear hypothesis.
+- [ ] Secondary metrics.
+- [ ] Counter metrics.
+- [ ] Predefined test duration, based on sample size.
+- [ ] A/B test code only includes eligible users that will be affected by the change.
+- [ ] QA checks on both control and test variants.
+
+**24-48 hours after launch:**
+
+- [ ] Volume of users assigned to each variant is as expected.
+- [ ] All logging is working correctly and in correct ratios.
+- [ ] No increase in crashes or errors.
+
+**At the end of the test duration**
+
+- [ ] Validate or invalidate your hypothesis based on experiment data.
+- [ ] Document your results and share with your team for feedback.
+- [ ] Ship the code for the winning variant. Delete the code for the losing variant.
```
I tried this out after you suggested it. It didn't look so great, so opted not to do it

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+---
+date: 2023-11-04
```
```suggestion
date: 2023-07-13
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+
+This is a good hypothesis because it enables you to break down which metrics you'll focus on in your test:
+
+- How many users watch the video.
+- What portion of the video did they view.
+- Successful interactions with the app features.
+- The number of customer support queries.
+- Churn rate.
```
```suggestion
- How many users watch the video
- What portion of the video did they view
- Successful interactions with the app features
- The number of customer support queries
- Churn rate
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+- How to implement and monitor your tests
+- How to analyze your results (statistically significant, or not)
+
+We'll also cover some frequently answered questions, such as when you _shouldn't_ run A/B tests, and we've added a simple [A/B test launch checklist]((#launch-checklist)) you can use when you run your own tests.
```
```suggestion
We'll also cover some frequently answered questions, such as when you _shouldn't_ run A/B tests, and we've added a simple [A/B test launch checklist](#a-b-test-launch-checklist) you can use when you run your own tests.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+
+3. **Your desired level of confidence.** The industry standard is 95%.
+
+You then use a [formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself – [A/B tests in PostHog](/ab-testing/features) calculate this for you automatically.
```
```suggestion
You then use a [sample size determination formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself – [A/B tests in PostHog](/ab-testing/features) calculate this for you automatically.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+
+3. **Any counter metrics.** These are metrics that ensure your user experience isn't deteriorating. For example, if your goal is to increase sign-ups, you should also verify your churn rate is not increasing.
+
+Most A/B testing tools, including [PostHog](/ab-testing/features), will calculate statistical significance for you. To do so manually, you'll need to input the number of users in each variant, and the conversion rate for each metric, into an [online calculator](https://abtestguide.com/calc/).
```
```suggestion
Most A/B testing tools, including [PostHog](/ab-testing/features), will calculate statistical significance for you. To do so manually, you'll need to input the number of users in each variant, and the conversion rate for each metric, into an online calculator.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
````diff
@@ -0,0 +1,262 @@
+Launching an A/B test requires careful planning to ensure accurate results and meaningful insights. To help you navigate this, we've put together this launch checklist:
+
+```
+**Before launch**
+
+- [ ] A clear goal metric.
+- [ ] A clear hypothesis.
+- [ ] Secondary metrics.
+- [ ] Counter metrics.
+- [ ] Predefined test duration, based on sample size.
+- [ ] A/B test code only includes eligible users that will be affected by the change.
+- [ ] QA checks on both control and test variants.
+
+**24-48 hours after launch:**
+
+- [ ] Volume of users assigned to each variant is as expected.
+- [ ] All logging is working correctly and in correct ratios.
+- [ ] No increase in crashes or errors.
+
+**At the end of the test duration**
+
+- [ ] Validate or invalidate your hypothesis based on experiment data.
+- [ ] Document your results and share with your team for feedback.
+- [ ] Ship the code for the winning variant. Delete the code for the losing variant.
````
```suggestion
**Before launch**

- [ ] A clear goal metric.
- [ ] A clear hypothesis.
- [ ] Secondary metrics.
- [ ] Counter metrics.
- [ ] Predefined test duration, based on sample size.
- [ ] Test code only includes eligible users that will be affected by change.
- [ ] QA checks on both control and test variants.

**24-48 hours after launch:**

- [ ] Volume of users assigned to each variant is as expected.
- [ ] All logging is working correctly and in correct ratios.
- [ ] No increase in crashes or errors.

**At the end of the test duration**

- [ ] Validate or invalidate your hypothesis based on experiment data.
- [ ] Document your results and share with your team for feedback.
- [ ] Ship code for the winning variant. Delete code for the losing variant.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+
+## How A/B testing works
+
+Feel free to skip this if you already understand the basics, but here's a quick ELI5 for newcomers.
```
```suggestion
Here's a quick ELI5 for newcomers – please skip if you're already familiar with the basics.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,262 @@
+No clear, measurable goal = ambiguous results you can't action.
+
+### 2. A clear hypothesis about why your changes will achieve your goal
+
+A well-defined hypothesis focuses your testing and guides your decision-making by providing a framework for evaluating your results. It should include your goal metric, how you think your change will improve it, and any other important context. For example:
```
```suggestion
Not having a well-defined hypothesis is one of the most [common A/B testing mistakes](/blog/ab-testing-mistakes) people make.

A good hypothesis focuses your testing and guides your decision-making by providing a framework for evaluating your results. It should include your goal metric, how you think your change will improve it, and any other important context. For example:
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+1. YOLO every change and hope they have the desired impact.
+2. Track user metrics and run A/B tests to verify your changes have the desired effect.
+
+One is easier than other, but the easy way rarely works in the long term. This guide is for software engineers who are just getting started with A/B testing. In it, you'll learn:
```
```suggestion
One is easier than the other, but the easy way rarely works in the long term. This guide is for software engineers who are just getting started with A/B testing. In it, you'll learn:
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+- [ ] Secondary metrics.
+- [ ] Counter metrics.
+- [ ] Predefined test duration, based on sample size.
+- [ ] Test code only includes eligible users that will be affected by change.
```
```suggestion
- [ ] Code only includes eligible users that will be affected by change.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+- [ ] Validate or invalidate your hypothesis based on experiment data.
+- [ ] Document your results and share with your team for feedback.
+- [ ] Ship code for the winning variant. Delete code for the losing variant.
```
```suggestion
- [ ] Ship winning variant code. Delete code for losing variant.
```

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+No clear, measurable goal = ambiguous results you can't action.
+
+### 2. A clear hypothesis about why your changes will achieve your goal
+Not having a well-defined hypothesis is one of the most [common A/B testing mistakes](/blog/ab-testing-mistakes) people make.
```
```suggestion
Not creating a well-defined hypothesis is one of the most [common A/B testing mistakes](/blog/ab-testing-mistakes).
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+- How to implement and monitor your tests
+- How to analyze your results (statistically significant, or not)
+
+We'll also cover some frequently answered questions, such as when you _shouldn't_ run A/B tests, and we've added a simple [A/B test launch checklist](#a-b-test-launch-checklist) you can use when you run your own tests.
```
```suggestion
We'll also cover some frequently asked questions, such as when you _shouldn't_ run A/B tests, and we've added a simple [A/B test launch checklist](#ab-test-launch-checklist) to use when you run your own tests.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+- **The test group** – These users will experience your product with your new changes.
+
+You then run your A/B test to gather data (usually a few days or weeks) and compare the differences in the goal metric between the different groups.
```
```suggestion
You then run your A/B test to gather data (usually over a few days or weeks) and compare the differences in the goal metric between the different groups.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+No clear, measurable goal = ambiguous results you can't action.
+
+### 2. A clear hypothesis about why your changes will achieve your goal
+Not creating a well-defined hypothesis is one of the most [common A/B testing mistakes](/blog/ab-testing-mistakes).
```
```suggestion

Not creating a well-defined hypothesis is one of the most [common A/B testing mistakes](/blog/ab-testing-mistakes).
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+The caveat here is that too small of change can slow your team down. Since running A/B tests take time, if you're constantly testing small changes, it will take longer to ship large, meaningful changes.
+
+A good rule of thumb to know if your change is too small is if it's unlikely to impact user behavior significantly. You can use qualitative data from user research or quantitative data from existing logging, or previous A/B tests, to inform this decision
```
```suggestion
A good rule of thumb to know if your change is too small is if it's unlikely to impact user behavior significantly. You can use qualitative data from user research, quantitative data from existing logging, or previous A/B tests to inform this decision
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+To know whether your sample size is large enough, you'll need to know:
+
+1. **The current conversion rate for your metric.** This is the percentage of users who performed the desired action (e.g., make a purchase, sign up) out of the total number of eligible users.
```
```suggestion
1. **The current count or conversion rate for your metric.** This is the count or percentage of users who performed the desired action (e.g., click a button, make a purchase, sign up) out of the total number of eligible users.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+1. **The current conversion rate for your metric.** This is the percentage of users who performed the desired action (e.g., make a purchase, sign up) out of the total number of eligible users.
+
+2. **Your "minimum detectable effect".** The smallest change in the conversion rate that you want to detect. The smaller the change, the larger the sample size you'll need.
```
```suggestion
2. **Your "minimum detectable effect".** The smallest change in the count or conversion rate that you want to detect. The smaller the change, the larger the sample size you'll need.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+* Monitor crashes or error rates to ensure that your test is not causing any issues. 
+
+Once you're sure that everything is running smoothly, try to avoid frequently checking the results until the test is finished. Checking too often can lead to reactive decisions based on incomplete data. It's usually best to wait until the test ends to make any decisions, as you'll have the best data at that point.
```
```suggestion
Once you're sure that everything is running smoothly, try to avoid frequently checking the results until the test is finished. Checking too often can lead to reactive decisions based on incomplete data. It's usually best to wait until the test ends to make any decisions, as you'll have more accurate data at that point.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+## Analyzing your results
+
+Once your test has run its duration, and you've collected enough data, you're ready to look at the results. What you're looking for is statistical significant change in any of the following metrics:
```
```suggestion
Once your test has run its duration and you've collected enough data, you're ready to look at the results. What you're looking for is a statistically significant change in any of the following metrics:
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+3. **Any counter metrics.** These are metrics that ensure your user experience isn't deteriorating. For example, if your goal is to increase sign-ups, you should also verify your churn rate is not increasing.
+
+Most A/B testing tools, including [PostHog](/ab-testing/features), will calculate statistical significance for you. To do so manually, you'll need to input the number of users in each variant, and the conversion rate for each metric, into an online calculator.
```
```suggestion
These metrics must have statistically significant changes because that means the impact is larger than random in 95% (or your confidence level) of cases. Most A/B testing tools, including [PostHog](/ab-testing), calculate statistical significance for you. To do so manually, you'll need to input the number of users in each variant, and the conversion rate for each metric, into an online calculator.
```

ivanagas replied:
I'm not sure if this is more or less confusing. It just might be useful to note why we care about statistical significant. Feel free to trash if it makes it worse.

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+Most A/B testing tools, including [PostHog](/ab-testing/features), will calculate statistical significance for you. To do so manually, you'll need to input the number of users in each variant, and the conversion rate for each metric, into an online calculator.
+
+### What if my results are _not_ statistically significant?
```
Good section 😄

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+- If you're considering a new caching strategy or database system, you could split your traffic between the old and new systems to see which performs better. 
+
+- When you're implementing a refactor to your codebase, you can use an A/B test to ensure your refactor does not perform worse.
```
```suggestion
- When you're refactoring your codebase, you can use an A/B test to ensure your refactor does not perform worse.
```

ivanagas commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,263 @@
+
+3. **Your desired level of confidence.** The industry standard is 95%.
+
+You then use a [sample size determination formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself – [A/B tests in PostHog](/ab-testing/features) calculate this for you automatically.
```
```suggestion
You then use a [sample size determination formula](https://en.wikipedia.org/wiki/Sample_size_determination) to determine if your sample size is large enough. There are many calculators online that will do this for you, so you can avoid calculating this yourself – [A/B tests in PostHog](/ab-testing) calculate this for you automatically.
```

ivanagas reviewed (approved):
Looks good, some minor cleanups

andyvan-ph commented on contents/blog/ab-testing-guide-for-engineers.md:
```diff
@@ -0,0 +1,264 @@
+---
+date: 2023-07-13
+title: "A software engineer's guide to A/B testing"
```
```suggestion
title: "Intro to A/B testing for software engineers"
```
